### ISSUE TITLE
To test STO where owners have reserved MSC

### ISSUE THREAD

**Issue by marv-engine** [OmniLayer/OmniJ#38]
Here's a way to test STO with owner(s) who have MSC that are reserved, e.g. because they have a tx20 sale (Dex phase 1) that's active.

Assign the same number of MSC, e.g. 100.0, to 3 addresses. Then:
1. Address 1 issues a tx20 for all 100.0 MSC, so his available balance is 0.0, reserved balance is 100.0
1. Address 2 issues a tx20 for 10.0 MSC, so his available balance is 90.0, reserved balance is 10.0
1. Address 3 doesn't do anything with his 100.0 MSC, so his available balance is 100.0, reserved balance is 0.0

The actor address acquires a number of MSC, e.g. 100.0 or more. Then the actor address submits an STO for all but 1.0 of his MSC. This will leave enough MSC available for the transfer fee for 100,000,000 recipient owners.

After the STO, check the balances of the 3 addresses. The (available balance + reserved balance) for all three should be exactly equal, with the available balance for each of them increased by the same amount.

**Comment by dexX7** [comment]
Hmm.. even if there are several rows in the test plan with "TDiv1" etc., for each test new addresses and properties are created, intended to reduce side effects and to have self sustaining tests. This works very well for tests involving properties, but unfortunally we can't simply reset the whole TMSC/MSC state on demand.

This is a fundamentally different approach than I used with [mastercore-rpc-tests](https://github.com/dexX7/mastercore-rpc-tests) where each test (or rather a test sequence) spawns at least one new, pristine node.

Long story short: I think it's given that we have no control over the MSC/TMSC state and therefore can't predict or command how many owners there are.

**Comment by marv-engine** [comment]
@dexX7 Sorry, I wasn't clear. For this test it doesn't matter how many other MSC owners there are or how many MSC they own. I'm just interested in seeing that the 3 newly created addresses all end up with exactly the same number of MSC after the STO - because before the STO those 3 addresses all have the same (available balance + reserved balance) total.

The 3 would end up with different amounts only if the number of MSC to be distributed runs out after 1 or 2 of the 3 get their distribution, which seems highly unlikely.

**Comment by dexX7** [comment]
@marv-engine: my mistake actually, I focused on the inability to control MSC holders and dismissed it without grasping the actual test. :)

I'll probably push it later.

**Comment by dexX7** [comment]
@marv-engine: is this one fine? https://github.com/dexX7/bitcoin-spock/commit/23e62e594d6ac75231b129ce7445b1cf7a09d503

As precaution: the test is skipped, if there are more owners than the actor can pay in fees.

A note regarding the test plans (CC: @msgilligan): while [sto-testplan.tsv](https://github.com/msgilligan/bitcoin-spock/blob/master/src/integ/groovy/foundation/omni/test/rpc/sto/sto-testplan.tsv) is completely parsed and used to generate tests, [sto-testplan-special.tsv](https://github.com/msgilligan/bitcoin-spock/blob/master/src/integ/groovy/foundation/omni/test/rpc/sto/sto-testplan-special.tsv) is not processed automatically, but hard coded, because the test class can't handle the unusual situations or constraints.

Tests in [MSCSendToOwnersTestPlanSpec.groovy](https://github.com/msgilligan/bitcoin-spock/blob/master/src/integ/groovy/foundation/omni/test/rpc/sto/MSCSendToOwnersTestPlanSpec.groovy) use [executeSendToOwners](https://github.com/msgilligan/bitcoin-spock/blob/master/src/integ/groovy/foundation/omni/test/rpc/sto/MSCSendToOwnersTestPlanSpec.groovy#L382-L398) which is overwritten in [MSCSendToOwnersTestPlanRawSpec.groovy](https://github.com/msgilligan/bitcoin-spock/blob/master/src/integ/groovy/foundation/omni/test/rpc/sto/MSCSendToOwnersTestPlanRawSpec.groovy) and each test (including the special ones) run two times: once by using the RPC call `sendtoowners_MP` and a second time by constructing a raw transaction (based on unobfuscated Omni data) via `sendrawtx_MP` to bypass any RPC checks, e.g. "property does not exist", to test the interpretation layer of Master/Omni Core.

**Comment by marv-engine** [comment]
@dexX7 - yes, your code for this test looks good to me.

On Sunday, February 1, 2015, dexX7 <notifications@github.com> wrote:

> @marv-engine <https://github.com/marv-engine>: is this one fine? dexX7@
> 23e62e5
> <https://github.com/dexX7/bitcoin-spock/commit/23e62e594d6ac75231b129ce7445b1cf7a09d503>
>
> As precaution: the test is skipped, if there are more owners than the
> actor can pay in fees.
>
> A note regarding the test plans (CC: @msgilligan
> <https://github.com/msgilligan>): while sto-testplan.tsv
> <https://github.com/msgilligan/bitcoin-spock/blob/master/src/integ/groovy/foundation/omni/test/rpc/sto/sto-testplan.tsv>
> is completely parsed and used to generate tests, sto-testplan-special.tsv
> <https://github.com/msgilligan/bitcoin-spock/blob/master/src/integ/groovy/foundation/omni/test/rpc/sto/sto-testplan-special.tsv>
> is not processed automatically, but hard coded, because the test class
> can't handle the unusual situations or constraints.
>
> Tests in MSCSendToOwnersTestPlanSpec.groovy
> <https://github.com/msgilligan/bitcoin-spock/blob/master/src/integ/groovy/foundation/omni/test/rpc/sto/MSCSendToOwnersTestPlanSpec.groovy>
> use executeSendToOwners
> <https://github.com/msgilligan/bitcoin-spock/blob/master/src/integ/groovy/foundation/omni/test/rpc/sto/MSCSendToOwnersTestPlanSpec.groovy#L382-L398>
> which is overwritten in MSCSendToOwnersTestPlanRawSpec.groovy
> <https://github.com/msgilligan/bitcoin-spock/blob/master/src/integ/groovy/foundation/omni/test/rpc/sto/MSCSendToOwnersTestPlanRawSpec.groovy>
> and each test (including the special ones) run two times: once by using the
> RPC call sendtoowners_MP and a second time by constructing a raw
> transaction (based on unobfuscated Omni data) via sendrawtx_MP to bypass
> any RPC checks, e.g. "property does not exist", to test the interpretation
> layer of Master/Omni Core.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/msgilligan/bitcoin-spock/issues/38#issuecomment-72363506>
> .
>


-- 
Marv Schneider
marv@engine.co

**Comment by dexX7** [comment]
@msgilligan: can be closed, I think.
